File: backend/app/schemas/report_subscription.py
"""Report-subscription schemas (scheduled PDF/Excel email delivery)."""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, field_validator


class SubscriptionCreate(BaseModel):
    recipient: str = Field(min_length=3, max_length=320)
    format: Literal["pdf", "xlsx"] = "pdf"
    schedule: Literal["hourly", "daily", "weekly"] = "daily"

    @field_validator("recipient")
    @classmethod
    def _valid_email(cls, v: str) -> str:
        v = v.strip()
        # Reject header-injection chars (CR/LF/comma/semicolon/space) BEFORE the
        # shape check so a value can never smuggle a Bcc:/CC: into the mail header.
        if any(c in v for c in "\r\n\t ,;") or "@" not in v or "." not in v.split("@")[-1]:
            raise ValueError("Düzgün email ünvanı deyil.")
        return v


class SubscriptionResponse(BaseModel):
    id: str
    saved_query_id: str
    recipient: str
    format: str
    schedule: str
    active: bool
    last_sent_at: str | None = None

File: frontend/src/api/reportSubscription.ts
import { client } from './client'

export type ReportFormat = 'pdf' | 'xlsx'
export type ReportSchedule = 'hourly' | 'daily' | 'weekly'

export interface Subscription {
  id: string
  saved_query_id: string
  recipient: string
  format: ReportFormat
  schedule: ReportSchedule
  active: boolean
  last_sent_at: string | null
}

export interface SubscriptionCreate {
  recipient: string
  format: ReportFormat
  schedule: ReportSchedule
}

export async function listSubscriptions(savedId: string): Promise<Subscription[]> {
  const { data } = await client.get<Subscription[]>(`/saved/${savedId}/subscriptions`)
  return data
}

export async function createSubscription(
  savedId: string,
  payload: SubscriptionCreate,
): Promise<Subscription> {
  const { data } = await client.post<Subscription>(`/saved/${savedId}/subscriptions`, payload)
  return data
}

export async function deleteSubscription(subId: string): Promise<void> {
  await client.delete(`/saved/subscriptions/${subId}`)
}
